### PR TITLE
[RDBMS] Hotfix: `az postgres flexible-server create` : Update VnetName to exclude servername and update default region for MySQL

### PR DESCRIPTION
**Pull request by arde0708**
This PR is a hotfix candidate for the recent Ignite Train. It includes :

- Changing Vnet Name and subnet name to exclude server name.( to allow servers to be provisioned with HA enabled )
- Change the MySQL default region to West US2 as MySQL Flexservers do not exist in the current default - EASTUS
- Update the connection string in the postgres create instance response.
- Skip adding special characters in the password
- Convert user supplied servername to lower case.

### CODE REVIEW

**Comment by yungezz** [comment]
hotfix candidate

**Comment by yungezz** [comment]
hi @rajsell @mjain2 @daeunyim could you pls review this PR? I cannot add you as reviewers,  it need to be in azure-sdk-read team I think

**Review by DaeunYim — APPROVED** [review]

**Review by mjain2 — APPROVED** [review]

**Review by rajsell — APPROVED** [review]

**Comment by mjain2** [comment]
> hi @rajsell @mjain2 @DaeunYim could you pls review this PR? I cannot add you as reviewers, it need to be in azure-sdk-read team I think

@yungezz we've signed off + checks looks green, thanks.

File: src/azure-cli/azure/cli/command_modules/rdbms/_flexible_server_util.py [hunk 1]
```diff
@@ -29,10 +30,13 @@ def create_random_resource_name(prefix='azure', length=15):
     return prefix + ''.join(digits)
 
 
-def generate_missing_parameters(cmd, location, resource_group_name, server_name):
+def generate_missing_parameters(cmd, location, resource_group_name, server_name, db_engine):
     # if location is not passed as a parameter or is missing from local context
     if location is None:
-        location = DEFAULT_LOCATION
+        if db_engine == 'postgres':
```
[the remark this answers is not in the document]
Comment (arde0708) [reply src/azure-cli/azure/cli/command_modules/rdbms/_flexible_server_util.py:36]: This is something that we are explicitly passing as lower case from the caller. Line 61 in flexible_server_custom_postgres.py.

[the remark this answers is not in the document]
Comment (yungezz) [reply src/azure-cli/azure/cli/command_modules/rdbms/_flexible_server_util.py:36]: could you pls add some test to change the overall change?

[the remark this answers is not in the document]
Comment (arde0708) [reply src/azure-cli/azure/cli/command_modules/rdbms/_flexible_server_util.py:36]: For this, we are just changing the default location from eastus to westus2 for MySQL.

**Review by yungezz — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/rdbms/_params.py [hunk 2]
```diff
@@ -240,9 +240,9 @@ def _flexible_server_params(command_group):
                 c.argument('zone', options_list=['--zone, -z'],
                            help='Availability zone into which to provision the resource.')
             elif command_group == 'mysql':
-                c.argument('tier', default='Burstable', validator=tier_validator,
+                c.argument('tier', default='Burstable',
                            help='Compute tier of the server. Accepted values: Burstable, GeneralPurpose, Memory Optimized ')
-                c.argument('sku_name', default='Standard_B1ms', options_list=['--sku-name'], validator=mysql_sku_name_validator,
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/rdbms/_params.py:245]: why removing validators here?

File: src/azure-cli/azure/cli/command_modules/rdbms/_params.py [hunk 3]
```diff
@@ -302,7 +300,9 @@ def _flexible_server_params(command_group):
                        help='Period of time (UTC) designated for maintenance. Examples: "Sun:23:30" to schedule on Sunday, 11:30pm UTC. To set back to default pass in "Disabled".')
             c.argument('tags', tags_type)
             if command_group == 'mysql':
-                c.argument('sku_name', options_list=['--sku-name'], validator=mysql_sku_name_validator,
+                c.argument('tier', options_list=['--tier'],
+                           help='Compute tier of the server. Accepted values: Burstable, GeneralPurpose, Memory Optimized ')
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/rdbms/_params.py:304]: For allowed values. prefer using get_enume_type()

File: src/azure-cli/azure/cli/command_modules/rdbms/_params.py [hunk 4]
```diff
@@ -302,7 +300,9 @@ def _flexible_server_params(command_group):
...
+                c.argument('sku_name', options_list=['--sku-name'],
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/rdbms/_params.py:305]: Same question: why removing validator?

File: src/azure-cli/azure/cli/command_modules/rdbms/flexible_server_custom_postgres.py [hunk 5]
```diff
@@ -59,7 +59,8 @@ def flexible_server_create(cmd, client,
 
         # Populate desired parameters
         location, resource_group_name, server_name = generate_missing_parameters(cmd, location, resource_group_name,
-                                                                                 server_name)
+                                                                                 server_name, 'postgres')
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/rdbms/flexible_server_custom_postgres.py:62]: please avoid positional arguments here

File: src/azure-cli/azure/cli/command_modules/rdbms/flexible_server_custom_postgres.py [hunk 6]
```diff
@@ -113,15 +114,15 @@ def flexible_server_create(cmd, client,
...
+        logger.warning('Make a note of your password. If you forget, you would have to '
+                       'reset your password with \'az postgres flexible-server update -n %s -g %s -p <new-password>\'.',
                        server_name, resource_group_name)
 
         _update_local_contexts(cmd, server_name, resource_group_name, location, user)
 
         return _form_response(user, sku, loc, server_id, host, version,
                               administrator_login_password if administrator_login_password is not None else '*****',
-                              _create_postgresql_connection_string(host, administrator_login_password), firewall_id,
+                              _create_postgresql_connection_string(host, user, administrator_login_password), firewall_id,
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/rdbms/flexible_server_custom_postgres.py:125]: avoid positional arguments here

File: src/azure-cli/azure/cli/command_modules/rdbms/_params.py [hunk 7]
```diff
@@ -240,9 +240,9 @@ def _flexible_server_params(command_group):
                 c.argument('zone', options_list=['--zone, -z'],
                            help='Availability zone into which to provision the resource.')
             elif command_group == 'mysql':
-                c.argument('tier', default='Burstable', validator=tier_validator,
+                c.argument('tier', default='Burstable',
                            help='Compute tier of the server. Accepted values: Burstable, GeneralPurpose, Memory Optimized ')
-                c.argument('sku_name', default='Standard_B1ms', options_list=['--sku-name'], validator=mysql_sku_name_validator,
+                c.argument('sku_name', default='Standard_B1ms', options_list=['--sku-name'],
                            help='The name of the compute SKU. Follows the convention Standard_{VM name}. Examples: Standard_B1ms, Standard_D4s_v3 ')
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/rdbms/_params.py:246]: Is it possible to get allowed sku values through SDK?

**Comment by Juliehzl** [comment]
HI @arde0708, as what I see in the e-mail, you need to fix the issue when creating a new flexible-server with HA in a VNET. May I know how you fix it?

File: src/azure-cli/azure/cli/command_modules/rdbms/_params.py (hunk 7 above)
Comment (arde0708) [reply src/azure-cli/azure/cli/command_modules/rdbms/_params.py:246]: > HI @arde0708, as what I see in the e-mail, you need to fix the issue when creating a new flexible-server with HA in a VNET. May I know how you fix it?

@Juliehzl : All those changes are in src/azure-cli/azure/cli/command_modules/rdbms/flexible_server_virtual_network.py. We are basically ensuring that the VNET and Subnet Name do not contain the servername to make this fix.

**Comment by arde0708** [comment]
> HI @arde0708, as what I see in the e-mail, you need to fix the issue when creating a new flexible-server with HA in a VNET. May I know how you fix it?

@Juliehzl All those changes are in src/azure-cli/azure/cli/command_modules/rdbms/flexible_server_virtual_network.py. Basically, we are removing the servername from the VNET name and subnet Name in all places to address this issue.

File: src/azure-cli/azure/cli/command_modules/rdbms/_params.py (hunk 2 above)
Comment (arde0708) [reply src/azure-cli/azure/cli/command_modules/rdbms/_params.py:245]: Addressed in the SDK comment.

File: src/azure-cli/azure/cli/command_modules/rdbms/_params.py (hunk 3 above)
Comment (arde0708) [reply src/azure-cli/azure/cli/command_modules/rdbms/_params.py:304]: Addressed in the SDK comment.

File: src/azure-cli/azure/cli/command_modules/rdbms/_params.py (hunk 4 above)
Comment (arde0708) [reply src/azure-cli/azure/cli/command_modules/rdbms/_params.py:305]: Addressed in the SDK comment.

File: src/azure-cli/azure/cli/command_modules/rdbms/_params.py (hunk 7 above)
Comment (arde0708) [reply src/azure-cli/azure/cli/command_modules/rdbms/_params.py:246]: We have a permanent fix for this in S176 where we fetch the SKU's from the SDK. We are currently removing the validators in the PR as there is a bug in the validator which prevents provisioning servers with certain SKU's

File: src/azure-cli/azure/cli/command_modules/rdbms/_params.py (hunk 3 above)
Comment (Juliehzl) [reply src/azure-cli/azure/cli/command_modules/rdbms/_params.py:304]: I mean all allowed values for tier and sku_name. Could we also get the allowed array from sdk?

File: src/azure-cli/azure/cli/command_modules/rdbms/_params.py (hunk 7 above)
Comment (Juliehzl) [reply src/azure-cli/azure/cli/command_modules/rdbms/_params.py:246]: sounds good

File: src/azure-cli/azure/cli/command_modules/rdbms/_params.py (hunk 3 above)
Comment (DaeunYim) [reply src/azure-cli/azure/cli/command_modules/rdbms/_params.py:304]: Yes we do have validators for tier, sku_name, storage size validated by the array from sdk. We will include that for the next train.

**Comment by Juliehzl** [comment]
> > HI @arde0708
> >             
> >             Aritra De
> >             FTE, as what I see in the e-mail, you need to fix the issue when creating a new flexible-server with HA in a VNET. May I know how you fix it?
> 
> @Juliehzl Zunli Hu FTE All those changes are in src/azure-cli/azure/cli/command_modules/rdbms/flexible_server_virtual_network.py. Basically, we are removing the servername from the VNET name and subnet Name in all places to address this issue.

Do you have test for related scenarios? I will recommend to add test cases to prevent potential issues in the future, including sdk changes.

**Comment by Juliehzl** [comment]
In addition, your command interface doesn't align with CLI command guideline. From the command, I have no idea which parameters are required.
image

See other command:
image

**Comment by arde0708** [comment]
> In addition, your command interface doesn't align with CLI command guideline. From the command, I have no idea which parameters are required.
> image
> 
> See other command:
> image

@Juliehzl Please note that none of the arguments in **az postgres flexible-server create** are **_Required_**. Hence, we dont have any of them marked as **_required_**.  If the user passes them, we use that. Else, we create a default one for the parameters not passed. as a result, you see the difference.

**Comment by yungezz** [comment]
I would suggest create an another issue to track the interface refinement work. Since this is a hotfix to mitigate immediate customer impact, better be more specific more efficient.

**Comment by Juliehzl** [comment]
Hi @arde0708, to do a quick hotfix, we will help you merge the PR.
But to avoid issues in the future, please add test for all your new commands and arguments. Especially there will be track2 migration in the future.
For command design, I will have some concern for custom easy usage, and we should also align with CLI command guideline.

**Review by Juliehzl — APPROVED** [review]

**Comment by arde0708** [comment]
> > > HI @arde0708
> > > ```
> > >         Aritra De
> > >         FTE, as what I see in the e-mail, you need to fix the issue when creating a new flexible-server with HA in a VNET. May I know how you fix it?
> > > ```
> > 
> > 
> > @Juliehzl Zunli Hu FTE All those changes are in src/azure-cli/azure/cli/command_modules/rdbms/flexible_server_virtual_network.py. Basically, we are removing the servername from the VNET name and subnet Name in all places to address this issue.
> 
> Do you have test for related scenarios? I will recommend to add test cases to prevent potential issues in the future, including sdk changes.

@Juliehzl : This is a temporary patch. The Backend Team is actually fixing the root cause of this issue, but it will be deployed later in October. So this is a patch just until then after which we will revert to what it was.

**Comment by Juliehzl** [comment]
Hi @arde0708, please make sure there is no breaking changes for existing commands and I will merge the PR for hotfix.
Here is to note that All CLI team will be on China National Day holiday from 10.1 ~ 10.8, and we could not support you quickly if there is any furthur issue.

**Pull request merged**